Add Fomu support

piotr-binkowski opened the pull request:
The aim of this PR is to add support for Fomu.
Initial design uses the Foboot configuration/firmware which requires special handling because it needs a custom BIOS, due to that at this point it is necessary to first run `make foboot-bios` to build the replacement BIOS before building other targets like `gateware`
It is also required to manually patch foboot with `firmware/foboot/putchar-fix.patch`

mithro commented on rtl/2-stage-1024-cache.v_toplevel_RegFilePlugin_regFile.bin:
```diff
@@ -0,0 +1,32 @@
+00000000000000000000000000000000
```
This CL shouldn't be adding an `rtl` directory. All `rtl` is under the `gateware` directory.

piotr-binkowski replied:
I've moved files from `rtl` directory to `gateware`

piotr-binkowski replied:
I've moved files from rtl directory to gateware

mithro commented on targets/fomu/Makefile.mk:
```diff
@@ -0,0 +1,79 @@
+
+# Image
+image-flash-$(PLATFORM):
+	iceprog $(IMAGE_FILE)
```
`iceprog` isn't use with Fomu -- you want `dfu-util`

piotr-binkowski replied:
At this point the `base` target for Fomu builds the whole bootloader image (gateware + foboot as bios). I think that loading it must be done using `iceprog` and a direct flash connection. Or is it possible to update the bootloader using `dfu-util`?

mithro replied:
It is possible to update the bootloader via dfu-util

piotr-binkowski commented:
@mithro please take a look
I think that I addressed all of your comments and if everything is all right then this can be merged

mithro commented on targets/fomu/base.py:
```diff
@@ -0,0 +1,451 @@
+        self.specials += AsyncResetSynchronizer(self.cd_por, self.reset)
+
+
+class SBLED(Module, AutoCSR):
```
This should be in `gateware` directory. I believe there is a version at https://github.com/timvideos/litex-buildenv/blob/master/gateware/ice40.py

mithro commented on targets/fomu/base.py:
```diff
@@ -0,0 +1,451 @@
+        )
+
+
+class SBWarmBoot(Module, AutoCSR):
```
Should be in https://github.com/timvideos/litex-buildenv/blob/master/gateware/ice40.py

mithro commented on targets/fomu/base.py:
```diff
@@ -0,0 +1,451 @@
+        ]
+
+
+class PicoRVSpi(Module, AutoCSR):
```
Should be in `gateware`.

mithro commented:
The aim here is to have a SoC you can load into the "user" part of the Fomu. This is the image that boots *after* the bootloader. See the "Migen and LiteX" section of https://workshop.fomu.im/#migen-and-litex

mithro commented:
Could you please rebase this pull request?

mithro commented:
@mateusz-holenko -- IIRC Could you work with @piotr-binkowski and @ewenmcneill to get this pull request merged?

piotr-binkowski commented:
I've rebased this and moved module definitions to `gateware` but I still need to update this so it generates a proper `user` image and not a bootloader.

piotr-binkowski commented:
Target now generates a normal user image that works without BIOS (assigning constant values to LEDs in gateware shows proper colors) so only thing missing now is booting BIOS or other software from Flash via XIP since BIOS doesn't fit inside BRAMs.

piotr-binkowski commented:
I've updated the target to use BIOS stored at 0x60000 in Flash (user gateware base address + gateware size) and I've modified the platform to by default target PVT variant. During testing I also noticed an issue with yosys version that is currently used but I will create an separate issue for that.
Please take a look @mithro
